File: ch3/3-6.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

void reverse(char s[]);
void itoa(int n, char s[], int m);

int main(void)
{
    int n = 324;
    int m = 5;
    char s[25];

    itoa(n, s, m);
    printf("%d = %s\n", n, s);
    return 0;
}

// reverse: reverse string s in place
void reverse(char s[])
{
    int c, i, j;
    for (i = 0, j = strlen(s)-1; i < j; i++, j--) {
        c = s[i];
        s[i] = s[j];
        s[j] = c;
    }
}

void itoa(int n, char s[], int m)
{
    int i, sign;
    sign = n;
    n = abs(n); // sets n to be positive
    i = 0;
    // generate digits in reverse order
    do {
        s[i++] = n % 10 + '0'; // get next digit
    } while ((n /= 10) > 0); // delete it
    if (sign < 0) {
        s[i++] = '-';
    }
    // add padding
    for (;i < m; i++) {
        s[i] = ' ';
    }
    s[i+1] = '\0';
    reverse(s);
}

File: ch3/3-6.go
package main

import (
	"fmt"
	"math"
)

const maxLen = 25

func main() {
	s1 := make([]byte, maxLen)
	n1 := 10
	s2 := make([]byte, maxLen)
	n2 := 955
	s3 := make([]byte, maxLen)
	n3 := math.MinInt64

	itoa(n1, s1, 10)
	itoa(n2, s2, 5)
	itoa(n3, s3, 8)
	fmt.Printf("itoa(%d) = %s\n", n1, string(s1))
	fmt.Printf("itoa(%d) = %s\n", n2, string(s2))
	fmt.Printf("itoa(%d) = %s\n", n3, string(s3))
}

func itoa(n int, s []byte, w int) {
	var i, sign int
	sign = n
	if sign < 0 {
		if n <= math.MinInt64 {
			n++
		}
		n = -n
	}
	i = 0
	for {
		s[i] = byte(n%10 + '0')
		i++
		n = n / 10
		if n == 0 {
			break
		}
	}
	if w > i {
		for ; i < w; i++ {
			s[i] = byte('0')
		}
	}
	if sign < 0 {
		s[i] = byte('-')
		i++
	}
	reverse(s, i)
}

func reverse(s []byte, l int) {
	var i, j int
	var c byte
	for j = l - 1; i < j; i, j = i+1, j-1 {
		c = s[i]
		s[i] = s[j]
		s[j] = c
	}
}
